File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/18-DFS/LeetCode/LeetCode-0112.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSumAux(self, root, targetSum, pathSum):
        if root is None: return False

        pathSum += root.val

        # Do the check when we are at a leaf node
        if root.left is None and root.right is None:
            return pathSum == targetSum

        if self.pathSumAux(root.left, targetSum, pathSum): return True
        return self.pathSumAux(root.right, targetSum, pathSum)
    
    
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        return self.pathSumAux(root, targetSum, 0)

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/18-DFS/LeetCode/LeetCode-0112.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool pathSumAux(TreeNode *root, int sum, int pathSum){
        if (root == NULL) return false;

        pathSum += root->val;

        // Do the check when we are at a leaf node
        if (root->left == NULL && root->right == NULL){
            if (pathSum == sum) return true;
            else                return false;
        } //end-if

        if (pathSumAux(root->left, sum, pathSum)) return true;
        return pathSumAux(root->right, sum, pathSum);
    } //end-pathSumAux

    bool hasPathSum(TreeNode* root, int sum) {
        if (root == NULL) return false;

        return pathSumAux(root, sum, 0);
    } //end-hasPathSum
};